src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/utils/CardinalityEstimation.java

```diff
@@ -0,0 +1,265 @@
+                cardinality = max(c, cardinality);
+            }
+        }
+        else if ( !possibleJoinVars.isEmpty() ){
+            for ( final Var v : possibleJoinVars ) {
+                int c1 = getCardinalityEstimationOfSpecificVar( pp1, v, mgr );
+                int c2 = getCardinalityEstimationOfSpecificVar( pp2, v, mgr );
+                int c = min(c1, c2);
+                cardinality = max(c, cardinality);
+            }
+        }
+        else {
+            for ( final Var v : allCertainVars ) {
+                int c ;
+                if ( pp1.getExpectedVariables().getCertainVariables().contains(v) ){
+                    c = getCardinalityEstimationOfSpecificVar( pp1, v, mgr );
+                }
+                else c = getCardinalityEstimationOfSpecificVar( pp2, v, mgr );
+                cardinality = max(c, cardinality);
+            }
+        }
+
+        return cardinality;
+    }
+
+    public int getCardinalityEstimationOfSpecificVar( final PhysicalPlan pp, final Var v, final FederationAccessManager mgr ) throws FederationAccessException {
+        if ( ppVarCardinalityTable.contains(pp, v) ){ return ppVarCardinalityTable.get(pp, v); }
+
+        final PhysicalOperatorForLogicalOperator pop = (PhysicalOperatorForLogicalOperator) pp.getRootOperator();
+        final LogicalOperator lop = pop.getLogicalOperator();
+
+        int cardinality = 0;
+        if (lop instanceof LogicalOpRequest) {
+            cardinality = getCardinalityEstimationOfLeafNode(pp, mgr);
+        }
+        else if (lop instanceof LogicalOpTPAdd) {
+            final PhysicalPlan pp1 = pp.getSubPlan(0);
+            final PhysicalPlan reqTP = formRequestBasedOnTPofTPAdd((LogicalOpTPAdd) lop);
+
+            return joinCardinalityBasedOnVar( pp1, reqTP, v, mgr );
+        }
+        else if (lop instanceof LogicalOpBGPAdd){
+            final PhysicalPlan pp1 = pp.getSubPlan(0);
+            final PhysicalPlan reqBGP = formRequestBasedOnBGPofBGPAdd((LogicalOpBGPAdd)lop);
+
+            return joinCardinalityBasedOnVar( pp1, reqBGP, v, mgr );
+        }
+        else if (lop instanceof LogicalOpJoin){
+            final PhysicalPlan pp1 = pp.getSubPlan(0);
+            final PhysicalPlan pp2 = pp.getSubPlan(1);
+
+            return joinCardinalityBasedOnVar( pp1, pp2, v, mgr );
+        }
+        else if (lop instanceof LogicalOpUnion){
+            final PhysicalPlan pp1 = pp.getSubPlan(0);
+            final PhysicalPlan pp2 = pp.getSubPlan(1);
+
+            final int c1 = getCardinalityEstimationOfSpecificVar( pp1, v, mgr );
+            final int c2 = getCardinalityEstimationOfSpecificVar( pp2, v, mgr );
+
+            cardinality = c1 + c2;
+        }
+
+        ppVarCardinalityTable.add(pp, v, cardinality);
+        return cardinality;
+    }
+
+    private int joinCardinalityBasedOnVar( PhysicalPlan pp1, PhysicalPlan pp2, Var v, FederationAccessManager mgr ) throws FederationAccessException {
```

Comment:
Replace 'private' by 'protected'. This way, potential subclasses can provide extended functionality.